src/Storages/StorageMerge.cpp

```diff
@@ -285,9 +285,16 @@ Pipe StorageMerge::read(
 
         if (processed_stage == QueryProcessingStage::FetchColumns && !storage_columns.getAliases().empty())
         {
+            ASTPtr where_expression;
+            if (query_info.query->as<ASTSelectQuery>()->where())
+                where_expression = query_info.query->as<ASTSelectQuery>()->where()->clone();
+
```

Comment:
So, the problem is, that aliases are replaced by theirs expression during query analysis, right? Maybe we have the same problem with other expressions?

Reply:
Looks, like we have. It still doesn't work if we query only alias column.

```sql
DROP TABLE IF EXISTS tt1;
DROP TABLE IF EXISTS tt2;
DROP TABLE IF EXISTS tt3;
DROP TABLE IF EXISTS tt_m;

CREATE TABLE tt1 (a UInt32, b UInt32 ALIAS a, c UInt32) ENGINE = Memory;
CREATE TABLE tt2 (a UInt32, b UInt32 ALIAS a * 2, c UInt32) ENGINE = Memory;
CREATE TABLE tt3 (a UInt32, b UInt32 ALIAS 0, c UInt32) ENGINE = Memory;
CREATE TABLE tt_m (a UInt32, b UInt32, c UInt32) ENGINE = Merge(currentDatabase(), 'tt1|tt2|tt3');

INSERT INTO tt1 (a, c) VALUES (1, 1);
INSERT INTO tt2 (a, c) VALUES (2, 2);
INSERT INTO tt3 (a, c) VALUES (3, 3);
```

Reply:
```sql
SELECT b  FROM tt_m WHERE b = 1

Received exception from server (version 21.12.1):
Code: 16. DB::Exception: Received from localhost:9000. DB::Exception: There is no column with name `b` in table default.tt3 (a4ee2161-e1a6-43aa-a4ee-2161e1a603aa). There are columns: a, c. (NO_SUCH_COLUMN_IN_TABLE)
```

```sql
SELECT b, count()  FROM tt_m GROUP BY b

Received exception from server (version 21.12.1):
Code: 16. DB::Exception: Received from localhost:9000. DB::Exception: There is no column with name `b` in table default.tt3 (a4ee2161-e1a6-43aa-a4ee-2161e1a603aa). There are columns: a, c. (NO_SUCH_COLUMN_IN_TABLE)
```

```sql
SELECT b FROM tt_m ORDER BY b

Received exception from server (version 21.12.1):
Code: 16. DB::Exception: Received from localhost:9000. DB::Exception: There is no column with name `b` in table default.tt3 (a4ee2161-e1a6-43aa-a4ee-2161e1a603aa). There are columns: a, c. (NO_SUCH_COLUMN_IN_TABLE)
```

```sql
SELECT b FROM tt_m LIMIT 1 BY b

Received exception from server (version 21.12.1):
Code: 16. DB::Exception: Received from localhost:9000. DB::Exception: There is no column with name `b` in table default.tt3 (a4ee2161-e1a6-43aa-a4ee-2161e1a603aa). There are columns: a, c. (NO_SUCH_COLUMN_IN_TABLE)
````

Reply:
If we add one more column (which doesn't participate in alias expression) some queries starts working, but some not.

```sql
SELECT c FROM tt_m WHERE b = 1

┌─c─┐
│ 1 │
└───┘
```

```sql
SELECT sum(c) FROM tt_m group by b

┌─sum(c)─┐
│      3 │
│      2 │
│      1 │
└────────┘
```

```sql
SELECT c FROM tt_m ORDER BY b

Received exception from server (version 21.12.1):
Code: 10. DB::Exception: Received from localhost:9000. DB::Exception: Not found column a in block. There are only columns: b, c. (NOT_FOUND_COLUMN_IN_BLOCK)
```

```sql
SELECT c FROM tt_m LIMIT 1 BY b

Received exception from server (version 21.12.1):
Code: 10. DB::Exception: Received from localhost:9000. DB::Exception: Not found column a in block. There are only columns: b, c. (NOT_FOUND_COLUMN_IN_BLOCK)
```

Reply:
Fixed